# A2: Vector Semantics

Nikolai Ilinykh, Mehdi Ghanimifard, Wafia Adouane and Simon Dobnik

The lab is an exploration and learning exercise to be done in a group and also in discussion with the teachers and other students.

Before starting, please read [the following instructions](https://github.com/sdobnik/computational-semantics/blob/master/README.md) on how to work on group assignments.

Write all your answers and the code in the appropriate boxes below.

In this lab we will look at how to build distributional semantic models from corpora and use semantic similarity captured by these models to do semantic tasks. We are also going to examine how different vector composition functions for phrases affect both the model and the learned information about similarities.  

Note that this lab uses a code from `dist_erk.py`, which contains functions that highly resemble those shown during the lecture. In the end, you can use either of the functions (from the lecture / from the file) to solve the tasks.

In [1]:
# the following command simply imports all the methods from that code.
from dist_erk import *

## 1. Loading a corpus

To train a distributional model, we first need a sufficiently large collection of texts which contain different words used frequently enough in different contexts. Here we will use a section of the Wikipedia corpus which you can download from [here](https://linux.dobnik.net/cloud/index.php/s/isMBj49jt5renYt?path=%2Fresources%2Fa2-distributional-representations) (wikipedia.txt.zip). (This file has been borrowed from another lab by [Richard Johansson](http://www.cse.chalmers.se/~richajo/)).  
When unpacked, the file is 151mb, hence if you are using the MLT servers you should store it in a temporary folder outside your home and adjust the `corpus_dir` path below.  
<!-- <It may already exist in `/opt/mlt/courses/cl2015/a5`.> -->


In [2]:
corpus_dir = '/scratch/lt2213-vt21/'

## 2. Building a model

Now you are ready to build the model.  
Using the methods from the code imported above build three word matrices with 1000 dimensions as follows:  

(i) with raw counts (saved to a variable `space_1k`);  
(ii) with PPMI (`ppmispace_1k`);  
(iii) with reduced dimensions SVD (`svdspace_1k`).  
For the latter use `svddim=5`. **[5 marks]**

Your task is to replace `...` with function calls. Functions are imported from `dist_erk.py` earlier, and they largely resemble functions shown during the lecture.

In [3]:
numdims = 1000
svddim = 5

"""
Possible function calls:
- do_word_count: words with counts, more frequent
- test_wordcount: a test with diff. dims (I don't think it's useful here I DON'T THINK TESTS ARE USEFUL HERE SO IMMA SKIP THEM)
--> make_word_index: index that maps words from 'keep_these_words'(from do_word_count) to their index
- co_occurrences: I think it makes pairs of words by context (?)
--> make_space: makes matrices of words with their contexts + counts
--> ppmi_transform: transform the space using positive pointwise mutual information (at least that's what the comments say)
--> svd_transform: transforming the space using singular value decomposition
- others but I don't think they're for this (???)
"""

# which words to use as targets and context words?
# we need to count the words and keep only the N most frequent ones
# which function would you use here with which variable?
ktw = do_word_count(corpus_dir, numdims)

wi = make_word_index(ktw) # word index
words_in_order = sorted(ktw) # sorted words

# create different spaces (the original matrix space, the ppmi space, the svd space)
# which functions with which arguments would you use here?
print('create count matrices')
space_1k = make_space(corpus_dir, wi, numdims) #def make_space(demo_dir, word_index, numdims)
print('ppmi transform')
ppmispace_1k = ppmi_transform(space_1k, wi) #def ppmi_transform(space, word_index)
print('svd transform')
svdspace_1k = svd_transform(space_1k, numdims, svddim) #def svd_transform(space, originalnumdimensions,keepnumdimensions)
print('done.')

reading file wikipedia.txt
create count matrices
reading file wikipedia.txt
ppmi transform
svd transform
done.


In [4]:
# now, to test the space, you can print vector representation for some words
print('house:', space_1k['house'])
print('time:', ppmispace_1k['time'])
print('summer:', svdspace_1k['summer'])

house: [2554 3774 3105  567  962  631  443  185  311  189  131   28   93  169
   81  125  151  408  194   90   79   29  217  184   62   15   31   70
   10    1   41   21    1   31   37    1   30    5   25    7    3   20
   11    1   32   36    2    5   66    4    0   46    8   18   28    0
   20    7    8   16   10   40    0  175   10    2    7   19    1  174
   11    3    1    6    0    0    0   10    9   11    7   24    4    4
   14   23   58    7    0   10    2    3   10    6   18    6   13    3
   22    0    3    5    3    7   14    3   40   20   19   15    6    8
   24    4    5    1   19    0    3    1    0   14    0   14   53    7
    7   11    6    5    5    4   12    6   53    1    1  433    4    0
    5    7    7   12    1    1    3    4   17    8   16    1    2   31
    1   12   14    1   44    6   14    9   38    7    2    6    8    1
   10    6   10    1    9    7    9    4    3   10    0   11    3    2
    0    2   11   37    2    0    2    1    5    9   10   16   88    6

Oxford Advanced Dictionary has 185,000 words, hence 1,000 words is not representative. We trained a model with 10,000 words, and 50 dimensions on truncated SVD. It took 40 minutes on a laptop. We saved all three matrices [here](https://linux.dobnik.net/cloud/index.php/s/isMBj49jt5renYt?path=%2Fresources%2Fa2-distributional-representations) (pretrained.zip). Download them and unpack them to a `pretrained` folder which should be a subfolder of the folder with this notebook:

In [5]:
import numpy as np

numdims = 10000
svddim = 50

print('Please wait...')
ktw_10k       = np.load('/scratch/lt2213-vt21/pretrained/ktw_wikipediaktw.npy', allow_pickle=True)
space_10k     = np.load('/scratch/lt2213-vt21/pretrained/raw_wikipediaktw.npy', allow_pickle=True).all()
ppmispace_10k = np.load('/scratch/lt2213-vt21/pretrained/ppmi_wikipediaktw.npy', allow_pickle=True).all()
svdspace_10k  = np.load('/scratch/lt2213-vt21/pretrained/svd50_wikipedia10k.npy', allow_pickle=True).all()
print('Done.')

Please wait...
Done.


In [6]:
# testing semantic space
print('house:', space_10k['house'])
print('time:', ppmispace_10k['time'])
print('summer:', svdspace_10k['summer'])

house: [2554 3774 3105 ...    0    0    0]
time: [0.         0.70427021 0.23319986 ... 0.         0.         0.        ]
summer: [-14.62484613   8.73395961  -6.58416212   0.88330785  -0.79893393
   2.32059622   2.40183216  -1.02680501   9.41382844  -4.82405627
  -5.18270272  -1.16056208   6.39978627   0.64449625   2.00152243
   0.82862145   0.05446969  -2.10563885  -1.09316833  -0.28541877
  -1.2432517   -1.31318143   2.0770079    0.02356217   2.3706459
   0.65020583   0.09939617   0.66939444   2.26833703  -0.03855381
  -2.96650461  -2.57877178  -2.22884732   1.2530818    3.2728387
   3.38833172   0.42561324   1.84383883   0.66728987   1.88300323
   3.7093252   -0.92727605   0.33814442  -1.8089893   -2.6320513
   2.19603751  -1.85283114   2.05096922  -1.87775569  -1.28438224]


## 3. Testing semantic similarity

The file `similarity_judgements.txt` (a copy is included with this notebook) contains 7,576 pairs of words and their lexical and visual similarities (based on the pictures) collected through crowd-sourcing using Mechanical Turk as described in [1]. The score range from 1 (highly dissimilar) to 5 (highly similar). Note: this is a different dataset from the phrase similarity dataset we discussed during the lecture (the one from [2]). For more information, please read the papers.

The following code will transform similarity scores into a Python-friendly format:

In [7]:
word_pairs = [] # test suit word pairs
semantic_similarity = [] 
visual_similarity = []
test_vocab = set()

for index, line in enumerate(open('similarity_judgements.txt')):
    data = line.strip().split('\t')
    if index > 0 and len(data) == 3:
        w1, w2 = tuple(data[0].split('#'))
        # it will check if both words from each pair exist in the word matrix.
        if w1 in ktw_10k and w2 in ktw_10k:
            word_pairs.append((w1, w2))
            test_vocab.update([w1, w2])
            semantic_similarity.append(float(data[1]))
            visual_similarity.append(float(data[2]))
        
print('number of available words to test:', len(test_vocab-(test_vocab-set(ktw))))
print('number of available word pairs to test:', len(word_pairs))
#list(zip(word_pairs, visual_similarity, semantic_similarity))

number of available words to test: 12
number of available word pairs to test: 774


Now we are going to test how the cosine similarity between vectors of each of the three spaces (normal space, ppmi, svd) compares with the human similarity judgements for the words in the similarity dataset. Which of the three spaces best approximates human judgements?

For comparison of several scores, we can use [Spearman correlation coefficient](https://en.wikipedia.org/wiki/Spearman's_rank_correlation_coefficient) which is implemented in `scipy.stats.spearmanr` [here](https://docs.scipy.org/doc/scipy-0.19.0/reference/generated/scipy.stats.spearmanr.html). The values of the Sperman correlation coefficient range from -1, 0 to 1, where 0 indicates no correlation, 1 perfect correaltion and -1 negative correlation. Hence, the greater the number the better the similarity scores align. The p values tells us if the coefficient is statistically significant. For this to be the case, it must be less than or equal to $< 0.05$.

Here is how you can calculate Pearson's correlation coefficient betweeen the scores of visual similarity and semantic similarity of the available words in the test suite:

In [8]:
from scipy import stats

rho, pval = stats.spearmanr(semantic_similarity, visual_similarity)
print("""Visual Similarity vs. Semantic Similarity:
rho     = {:.4f}
p-value = {:.4f}""".format(rho, pval))

Visual Similarity vs. Semantic Similarity:
rho     = 0.7122
p-value = 0.0000


Let's now calculate the cosine similarity scores of all word pairs in an ordered list using all three matrices. **[6 marks]**

In [9]:
from scipy import spatial

raw_similarities  = [1 - spatial.distance.cosine(space_10k[w1], space_10k[w2]) for w1, w2 in word_pairs]
ppmi_similarities = [1 - spatial.distance.cosine(ppmispace_10k[w1], ppmispace_10k[w2]) for w1, w2 in word_pairs]
svd_similarities  = [1 - spatial.distance.cosine(svdspace_10k[w1], svdspace_10k[w2]) for w1, w2 in word_pairs]

# print(raw_similarities)

Now, calculate correlation coefficients between lists of similarity scores and the real semantic similarity scores from the experiment. The scores of what model best correlates them? Is this expected? **[6 marks]**

In [10]:
# rho, pval = stats.spearmanr(semantic_similarity, visual_similarity)
# print("""Visual Similarity vs. Semantic Similarity:
# rho     = {:.4f}
# p-value = {:.4f}""".format(rho, pval))

# what are the real semantic similarity scores tho???????????????? i'll assume it's semantic_similarity but???????????
# i mostly copied from 2 cells above so idk if that's right?????????

# raw similarity
rho_raw_sim, pval_raw_sim = stats.spearmanr(semantic_similarity, raw_similarities)
print("""Semantic Similarity vs. Raw Similarity:
rho     = {:.4f}
p-value = {:.4f}""".format(rho_raw_sim, pval_raw_sim))

# ppmi similarity
rho_ppmi_sim, pval_ppmi_sim = stats.spearmanr(semantic_similarity, ppmi_similarities)
print("""Semantic Similarity vs. Ppmi Similarity:
rho     = {:.4f}
p-value = {:.4f}""".format(rho_ppmi_sim, pval_ppmi_sim))

# Svd similarity
rho_svd_sim, pval_svd_sim = stats.spearmanr(semantic_similarity, svd_similarities)
print("""Visual Similarity vs. Semantic Similarity:
rho     = {:.4f}
p-value = {:.4f}""".format(rho_svd_sim, pval_svd_sim))

Semantic Similarity vs. Raw Similarity:
rho     = 0.1522
p-value = 0.0000
Semantic Similarity vs. Ppmi Similarity:
rho     = 0.4547
p-value = 0.0000
Visual Similarity vs. Semantic Similarity:
rho     = 0.4232
p-value = 0.0000


**Your answer should go here:** The model that best correlates is the PPMI one (with a value of 0.4547). The reason behind this is that PPMI (Positive Pointwise Mutual Information) considers context words that are particularly informative, so it makes sense (I *think*) that its score is a lot higher than that of raw similarity.

We can also calculate correlation coefficients between lists of cosine similarity scores and the real visual similarity scores from the experiment. Which similarity model best correlates with them? How do the correlation coefficients compare with those from the previous comparison - and can you speculate why do we get such results? **[7 marks]**

In [11]:
# is this the exact same thing as before but with visual_similarity instead of semantic_similarity???????????
# asfjskfjskjasj

# raw similarity
rho_raw_sim_v2, pval_raw_sim_v2 = stats.spearmanr(visual_similarity, raw_similarities)
print("""Semantic Similarity vs. Raw Similarity:
rho     = {:.4f}
p-value = {:.4f}""".format(rho_raw_sim_v2, pval_raw_sim_v2))

# ppmi similarity
rho_ppmi_sim_v2, pval_ppmi_sim_v2 = stats.spearmanr(visual_similarity, ppmi_similarities)
print("""Semantic Similarity vs. Ppmi Similarity:
rho     = {:.4f}
p-value = {:.4f}""".format(rho_ppmi_sim_v2, pval_ppmi_sim_v2))

# Svd similarity
rho_svd_sim_v2, pval_svd_sim_v2 = stats.spearmanr(visual_similarity, svd_similarities)
print("""Visual Similarity vs. Semantic Similarity:
rho     = {:.4f}
p-value = {:.4f}""".format(rho_svd_sim_v2, pval_svd_sim_v2))

Semantic Similarity vs. Raw Similarity:
rho     = 0.1212
p-value = 0.0007
Semantic Similarity vs. Ppmi Similarity:
rho     = 0.3838
p-value = 0.0000
Visual Similarity vs. Semantic Similarity:
rho     = 0.3097
p-value = 0.0000


**Your answer should go here:** The PPMI model stays the highest, but in general all scores are a bit lower. But this time Raw similarity's p-value gives us a number other than 0, so that's something???? (Idk what that means exactly tbh).
Comparing these scores to the previous ones, it's obvious that the reason why the ones with visual similarity are lower is because the values in similarity_judgements.txt are generally lower for visual similarity than for semantic similarity. Why's that? No clue. Does anyone know where this data comes from? :/

## 4. Operations on similarities

We can perform mathematical operations on vectors to derive meaning predictions. For example, we can subtract the normalised vectors for `king` minus `queen` and add the resulting vector to `man` and we hope to get the vector for `woman`. Why? **[3 marks]**

**Your answer should go here:** We are able to get analogies from the relations between two words (analogy = 'a is to b what c is to d"). In this case, we could say that "a man is to a woman what a king is to a queen", because man-woman share a male-female relationship in the same way as king-queen do.

We have the following 4 assumptions:
- the group of all 'person' entities are either 'man' or 'woman', so: person = {man, woman} (let's ignore non-binary people here for a second, I'm sorry)
- the group of all 'monarch' entities are either 'king' or 'queen', so: monarch = {king, queen}
- both 'woman' and 'queen' belong to the same group 'female', so: female = {woman, queen}
- both 'man' and 'king' belong to the same group 'male', so: male = {man, king}

With these, we can deduce that:
- king - man + woman = queen (we get the property 'male' away from king and add the property 'female', so it becomes queen)
- queen - woman + man = king (we get the property 'female' away from queen and add the property 'male', so it becomes king)
- man - king + queen = woman (we get the property 'monarch' away from man and add the property 'female', so it becomes woman)
- woman - queen + king = man (we get the property 'monarch' away from woman and add the property 'male', so it becomes man)

(if the teachers are looking for a more math-y explanation, I don't have it)

Here is some helpful code that allows us to calculate such comparisons.

In [12]:
from scipy.spatial import distance

def normalize(vec):
    return vec / veclen(vec)

def find_similar_to(vec1, space):
    # vector similarity funciton
    #sim_fn = lambda a, b: 1-distance.euclidean(normalize(a), normalize(b))
    #sim_fn = lambda a, b: 1-distance.correlation(a, b)
    #sim_fn = lambda a, b: 1-distance.cityblock(normalize(a), normalize(b))
    #sim_fn = lambda a, b: 1-distance.chebyshev(normalize(a), normalize(b))
    #sim_fn = lambda a, b: np.dot(normalize(a), normalize(b))
    sim_fn = lambda a, b: 1-distance.cosine(a, b)

    sims = [
        (word2, sim_fn(vec1, space[word2]))
        for word2 in space.keys()
    ]
    return sorted(sims, key = lambda p:p[1], reverse=True)

Here is how you apply this code. Comment on the results you get. **[3 marks]**

In [13]:
short = normalize(svdspace_10k['short'])
light = normalize(svdspace_10k['light'])
long = normalize(svdspace_10k['long'])
heavy = normalize(svdspace_10k['heavy'])

find_similar_to(light - (heavy - long), svdspace_10k)[:10]

[('long', 0.8733111261346901),
 ('above', 0.8259671977311955),
 ('around', 0.8030776291120685),
 ('sun', 0.7692439111243973),
 ('just', 0.7678481974778111),
 ('wide', 0.767257431992253),
 ('each', 0.7665960260861158),
 ('circle', 0.7647746702909336),
 ('length', 0.7601066921319761),
 ('almost', 0.7542351860536628)]

**Your answer should go here:** It's trying to find words similar to 'short' (I guess, although it's never specified, but if we have 'light' and from that we subtract 'heavy' and 'long' I guess we kinda get 'short'??). We first get 'long' (an antonym), then 'above' (kind of a word more closely related to the antonym than to 'short', maybe it should be 'under'), then 'around' (no idea how it relates to 'short'), then sun (no idea again, how can the sun be short? maybe bc of texts talking about distances among the sun and Mercury for example??), then just (maybe bc of the collocation 'just short of'??), then wide (again, kind of an antonym), then each (no clue here), then circle (hmmm a short circle? circle is not a shape that can be short to me), then length (this makes sense bc 'short' is heavily related to length), then almost (this also makes sense, like almost getting there but oh no, too short or sth like that).

Find 5 similar pairs of pairs of words and test them. Hint: Google for `word analogies examples`. You can also construct analogies that are less lexical but more grammatical, e.g. `see, saw, leave, ?` or analogies that are based on world knowledge as in the [Google analogy dataset](http://download.tensorflow.org/data/questions-words.txt) from [3]. Does the resulting vector similarity confirm your expectations? But remember you can only do this if the words are contained in our vector space with 10,000 dimensions. **[10 marks]**

In [60]:
# 1 short shortest cold coldest
small = normalize(svdspace_10k['small'])
smallest = normalize(svdspace_10k['smallest'])
old = normalize(svdspace_10k['old'])
oldest = normalize(svdspace_10k['oldest'])
# try to find short
print('Words similar to \'short\'')
print(find_similar_to(smallest - (oldest - old), svdspace_10k)[:10])

# 2 Dallas Texas Miami Florida
dallas = normalize(svdspace_10k['dallas'])
texas = normalize(svdspace_10k['texas'])
miami = normalize(svdspace_10k['miami'])
florida = normalize(svdspace_10k['florida'])
# try to find dallas
print('Words similar to \'Dallas\'')
print(find_similar_to(texas - (florida - miami), svdspace_10k)[:10])

# 3 duck bird ray fish
duck = normalize(svdspace_10k['duck'])
bird = normalize(svdspace_10k['bird'])
ray = normalize(svdspace_10k['ray']) #obviously BAD choice of fish bc then I get men's names oops BUT MOST FISH NAMES WEREN'T AVAILABLE I'M SORRY
fish = normalize(svdspace_10k['fish'])
# try to find duck
print('Words similar to \'duck\'')
print(find_similar_to(bird - (fish - ray), svdspace_10k)[:10])

# 4 mango fruit pepper vegetable
mango = normalize(svdspace_10k['apple'])
fruit = normalize(svdspace_10k['fruit'])
lettuce = normalize(svdspace_10k['pepper'])
vegetable = normalize(svdspace_10k['vegetable'])
# try to find mango
print('Words similar to \'mango\'')
print(find_similar_to(fruit - (vegetable - lettuce), svdspace_10k)[:10])

# 5 Europe euro Canada dollar
europe = normalize(svdspace_10k['europe'])
euro = normalize(svdspace_10k['euro'])
canada = normalize(svdspace_10k['canada'])
dollar = normalize(svdspace_10k['dollar'])
# try to find europe
print('Words similar to \'Europe\'')
print(find_similar_to(euro - (dollar - canada), svdspace_10k)[:10])

Words similar to 'short'
[('smallest', 0.7785521034741371), ('above', 0.7649074515990197), ('hence', 0.7610292280555147), ('exactly', 0.7500413591976072), ('equivalent', 0.7485218338606945), ('same', 0.7410230371285738), ('empty', 0.7397074042720176), ('zero', 0.7266534320068492), ('one', 0.7217772688736775), ('metric', 0.7147120820073954)]
Words similar to 'Dallas'
[('park', 0.8659902072270402), ('heights', 0.8617585796820914), ('texas', 0.8600942301388608), ('grove', 0.850446820641441), ('located', 0.8503715772887688), ('xxx', 0.8492631413085925), ('township', 0.8467369106093091), ('cemetery', 0.8420342516626697), ('oklahoma', 0.8370803529007436), ('ontario', 0.8306341914418919)]
Words similar to 'duck'
[('ray', 0.883630818956996), ('charlie', 0.8464977386005359), ('pete', 0.8331306334286213), ('johnny', 0.8328410108145037), ('ben', 0.8302722928085775), ('bob', 0.8263225192135651), ('tom', 0.8241111326855491), ('eric', 0.8235367818062104), ('alex', 0.822594973991159), ('ted', 0.82194

## 5. Semantic composition and phrase similarity **[20 marks]**

In this task, we are going to look at how different semantic composition models, introduced in [2] correlate with human judgements. The file with the dataset is `mitchell_lapata_acl08.txt` included with this notebook. Your task is to do the following:  

(i) process the dataset, extract pairs of `reference - landmark high` and `reference - landmark low`; you can use the code from the lecture as something to start with. Note that there are 2 landmarks for each reference: one landmark exhibits high similarity with the reference, while another one has low similarity with the reference. A single human participant could have evaluated both of these pairs. For more details, we refer you to the paper.  

(ii) build models of semantic phrase composition: in the lecture we introduced simple additive, simple multiplicative and combined models (details are in [2]). Your task is to take a single pair (a reference or a high similarity landmark or a low similarity landmark) and compute the composition of its vectors using each of these functions. Thus, you will have three compositional models that take a `noun - verb` pair and output a single vector, representing the meaning of this pair. As your semantic space, you can use pretrained spaces (standard space, ppmi or svd) introduced above. It is up to you which space you use, but for someone who runs your code, it should be pretty straightforward to switch between them.

(iii) calculate Spearman correlation between each model's predictions and human judgements; you should have something similar to the scores that are shown in the paper [2]:  

![title](./res.png)

The paper states that they calculated correlations between each individual participant's judgeements and each model's predictions.  

Let's say we have 3 models: simple additive (A), simple multiplicative (M), combined (C).  
From our task dataset, we also know that we have 20 participants.  
Now, for each participant in 20 participants we get all `verb - noun` pairs that these participated evaluated.  
For example:

In [ ]:
participant_judgemenets_example = [
 'participant50 chatter child gabble 6 high',
 'participant50 chatter tooth click 2 high',
 'participant50 reel head whirl 5 high',
 'participant50 reel mind stagger 4 low',
 'participant50 reel industry stagger 5 high',
 'participant50 reel man whirl 3 low',
 'participant50 glow fire beam 7 low',
 'participant50 glow face burn 3 low',
 'participant50 glow cigar burn 5 high',
 'participant50 glow skin beam 7 high'
    
]

In [ ]:
participant_judgemenets_example

Let's look at the first pair that participant50 evaluated: reference `child chatter` and high-level similarity landmark (as the last word in the row indicates) `child gabble`. The human gave the similarity score of 6 (very similar). Thus, human similarity judgment = [6].  

Our A model's output:  
cosine(p1, p2) = 0.88, where p1 is the result of addition of word vectors in the reference phrase `child gabble`, and p2 is the result of addition of word vectors in the high-level similarity phrase `child chatter`.  

Therefore, we have human rating vector [6] and model A output [0.88]. Next is to compute correlation between these two vectors.

To get an overall score, simply average your correlation scores over all participants, since you are calculating correlation scores per participant.

Of course, your human rating vectors will be longer (e.g., [6, 7, 3, 4, 5]) where each element is a participant's judgement of a specific pair. Each of your models (A, B, C) will produce a single vector of cosine similarity between these same pairs (e.g., [0.89, 0.98, 0.23, 0.65, 0.55]). The goal is to compare each model's cosine similarity vectors with human rating vectors and identify the model which outputs the best result in terms of being the closest to the way human rate similarity between the phrases.

The minimum to do in this task: compute correlations for 3 models mentioned above and human rating for AT LEAST one participant. Elaborate on how different the resulting correlation scores are depending on the model's composition function (additive, multiplicative, combined). For examples on how to interpret the results, look at Section 5 Results of the original paper.

In [ ]:
#your code should go here

**Any comments/thoughts should go here:**

# Literature

  - [1] C. Silberer and M. Lapata. Learning grounded meaning representations with autoencoders. In Proceedings of the 52nd Annual Meeting of the Association for Computational Linguistics, pages 721–732, Baltimore, Maryland, USA, June 23–25 2014 2014. Association for Computational Linguistics.  

  - [2] Mitchell, J., & Lapata, M. (2008). Vector-based Models of Semantic Composition. In Proceedings of ACL-08: HLT (pp. 236–244). Association for Computational Linguistics.
  
  - [3] T. Mikolov, I. Sutskever, K. Chen, G. S. Corrado, and J. Dean. Distributed representations of words and phrases and their compositionality. In Advances in neural information processing systems, pages 3111–3119, 2013.

## Marks

This assignment has a total of 60 marks.